crates/red_knot/src/salsa_db/semantic.rs

```diff
@@ -0,0 +1,577 @@
+        fn enter_node(&mut self, node: AnyNodeRef) -> TraversalSignal {
+            // Don't traverse into expressions
+            if node.is_expression() {
+                return TraversalSignal::Skip;
+            }
+
+            TraversalSignal::Traverse
+        }
+
+        fn visit_stmt(&mut self, stmt: &Stmt) {
+            match stmt {
+                Stmt::Import(import) => {
+                    for alias in &import.names {
+                        self.dependencies.push(ModuleName::new(&alias.name));
+                    }
+                }
+
+                Stmt::ImportFrom(from) => {
+                    if let Some(level) = NonZeroU32::new(from.level) {
+                        // FIXME how to handle dependencies if the current file isn't a module?
+                        //  e.g. what if the current file is a jupyter notebook. We should still be able to resolve files somehow.
+                        if let Some(resolved_module) = &self.resolved_module {
+                            if let Some(dependency) = resolved_module.resolve_dependency(
+                                self.db,
+                                &Dependency::Relative {
+                                    module: from
+                                        .module
+                                        .as_ref()
+                                        .map(|module| ModuleName::new(module)),
+                                    level,
+                                },
+                            ) {
+                                self.dependencies.push(dependency);
+                            };
+                        }
+                    } else {
+                        let module = from.module.as_ref().unwrap();
+                        self.dependencies.push(ModuleName::new(module));
+                    }
+                }
+                _ => {}
+            }
+            preorder::walk_stmt(self, stmt);
+        }
+    }
+
+    let parsed = source::parse(db.upcast(), file);
+
+    let mut visitor = DependenciesVisitor {
+        db,
+        resolved_module: file_to_module(db, file),
+        dependencies: Vec::new(),
+    };
+
+    // TODO change the visitor so that `visit_mod` accepts a `ModRef` node that we can construct from module.
+    visitor.visit_body(&parsed.syntax().body);
+
+    Arc::from(visitor.dependencies)
+
+    // TODO we should extract the names of dependencies during parsing to avoid an extra traversal here.
+}
+
+#[salsa::jar(db=Db)]
+pub struct Jar(
+    ModuleSearchPaths,
+    FunctionTypingScope,
+    ClassTypingScope,
+    Module,
+    ast_ids::ast_ids,
+    semantic_index,
+    symbol_table_query,
+    flow_graph::flow_graph,
+    dependencies,
+    resolve_module,
+    file_to_module,
+    typing_scopes,
+    infer_function_body,
+    infer_class_body,
+    infer_module_body,
+);
+
+pub trait Db: source::Db + salsa::DbWithJar<Jar> + Upcast<dyn source::Db> {}
```

Comment:
To me these are kind of the least important types because it's mainly just boilerplate to make Salsa happy. However, I also expect that we'll probably have a `db` module in every Salsa-enabled crate rather than smashing all of this together into a single file. The way it is in this PR is just for fast prototyping.